### ISSUE TITLE
Unit tests don't all pass, due to HTTP 429 throttling

### ISSUE THREAD

**Issue by rbcarson** [Azure/azure-documentdb-python#30]
When I run the unit tests, many of them end up failing due to throttling:

```
Error
Traceback (most recent call last):
  File "C:\Users\clam\Documents\azure-documentdb-python\test\crud_tests.py", line 1758, in test_create_default_indexing_policy_name_based
    self._test_create_default_indexing_policy(True);
  File "C:\Users\clam\Documents\azure-documentdb-python\test\crud_tests.py", line 1797, in _test_create_default_indexing_policy
    'path': '/*'
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\document_client.py", line 222, in CreateCollection
    options)
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\document_client.py", line 1719, in Create
    headers)
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\document_client.py", line 1878, in __Post
    headers=headers)
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\synchronized_request.py", line 162, in SynchronizedRequest
    return _InternalRequest(connection_policy, request_options, request_body)
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\synchronized_request.py", line 99, in _InternalRequest
    raise errors.HTTPFailure(response.status, data, headers)
HTTPFailure: Status code: 429
{"code":"429","message":"Message: {\"Errors\":[\"Request rate is large\"]}\r\nActivityId: 4ff91177-2234-45ab-b5c6-65668ea15a14, Request URI: /apps/28284a68-9708-4ee8-97da-dc5bfad115b9/services/66452474-7c6b-45c8-b2fc-6df0e4f4b5ce/partitions/b36e3fc6-a803-4fb9-893f-1572dd11022f/replicas/130915761969947884p"}
```

Sometimes the server response is slightly different:
```
Traceback (most recent call last):
  File "C:\Users\clam\Documents\azure-documentdb-python\test\crud_tests.py", line 394, in test_spatial_index_name_based
    self._test_spatial_index(True);
  File "C:\Users\clam\Documents\azure-documentdb-python\test\crud_tests.py", line 416, in _test_spatial_index
    'path': '/'
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\document_client.py", line 222, in CreateCollection
    options)
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\document_client.py", line 1719, in Create
    headers)
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\document_client.py", line 1878, in __Post
    headers=headers)
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\synchronized_request.py", line 162, in SynchronizedRequest
    return _InternalRequest(connection_policy, request_options, request_body)
  File "C:\Users\clam\Documents\azure-documentdb-python\pydocumentdb\synchronized_request.py", line 99, in _InternalRequest
    raise errors.HTTPFailure(response.status, data, headers)
HTTPFailure: Status code: 429
{"code":"429","message":"The request rate is too large. Please retry after sometime.\r\nActivityId: 75450dbb-3eef-48fc-82c9-df8df7c0dede"}
```

I tried adding a delay of 15 seconds between tests, but this only reduces the number of failures. I'm not the owner of the DocumentDB account, so I don't know what performance level is configured, or what kind of load the Collection is experiencing. It's a testing account that _should_ be seeing little or no other usage.

**Comment by rnagpal** [comment]
Hi Carson,

Thanks for reporting this issue.

Running these tests all at the same time might result in 429s as you see but I'm surprised that adding the delay of 15 secs between the tests is still causing the throttle. In the second case, are the same set of test cases failing? Probably, each of those tests are making several requests too soon causing the throttle within the same test case. Most of the collections in the tests should be S1 which support 250 RUs/second and I'd guess that would be sufficient for most of the test cases.

Performance tier is tied to a collection, so as long as you are creating new test collections there shouldn't be any impact of the other load(if any) on the same database.

We probably would need to add retry policy around 429s for these test cases, so that always succeed(even if they take time).

I'll add this to our backlog but feel free to contribute to it if you wish.

Regards,
Rajesh

**Comment by rbcarson** [comment]
Hi Rajesh. Thanks for the insight. I believe that the failures with the 15 second delay are a subset of the failures with no delay -- I didn't look too closely at the results because without any delay, about half the tests fail.

It would be great to have a retry policy for all the methods in pydocumentdb. I am open to contributing, but I'm not yet familiar enough with DocumentDB. [This post](https://azure.microsoft.com/en-us/blog/performance-tips-for-azure-documentdb-part-2/) mentions how the .NET API has built-in retry for 429 responses, and I see signs of this being implemented in some parts of pydocumentdb too. I may file another issue about this tomorrow, because I have some suspicions about whether it's working correctly.

**Comment by rnagpal** [comment]
The 2 slightly different versions of the error messages you were seeing for throttle scenarios was due to the fact that sometimes throttle is detected at the gateway level and Gateway sends RequestTooLargerException(the second one that you are seeing). The first one is coming directly from the server and hence has a slightly different message but since both of them have status code 429, they are due to the same reason.

It will be a while until we implement the throttle retry policy for all documentdb APIs for Python SDK but you can take a look at http://stackoverflow.com/questions/29275447/exception-microsoft-azure-documents-requestratetoolargeexception-while-querying as an example how that's implement in .NET SDK, so you may have to do similar thing for Python. In short, you catch the exception, check if it's status code is 429, get the value of "RetryAfterInMilliseconds" property from the response header and wait until that many ms, and resubmit the request. You can loop for say 3 times before giving up.

Meanwhile, we should be able to modify the tests so that each one of them doesn't creates a new collection every time. We can have one or two test collections created which are shared by most of the tests for creating resources within that collection and cleaning them up at the end. Collection creates are the only one which are causing problems in this case, so if we can eliminate that we should be good. We may be able to get to this approach in our next release but if you are comfortable enough to make these changes, feel free to do so and we will accept it as a PR.

Regards,
Rajesh

**Comment by rnagpal** [comment]
Looks like there is something already present currently: https://github.com/Azure/azure-documentdb-python/blob/master/pydocumentdb/backoff_retry_utility.py

**Comment by rbcarson** [comment]
For now, I'm experimenting with the REST API without using pydocumentdb. I'll try implementing automatic retry based on your link - thanks.

**Comment by rnagpal** [comment]
@rbcarson : This issue has been fixed with 1.9.0 release. You should be able to run the tests without any custom retry logic.
